### PR TITLE
feat: encode preview in parallel

### PR DESCRIPTION
**Pull request by anubhavitis**
## Enhanced Capture Mechanism:

Now, the capture process followed these steps:

- Creation of a Frame channel before the recorder initiated recording.
- A new thread is spawned to receive Frames from those channel and encode them.
- The primary thread is responsible for recording frames and saving them in the state for future GIF encoding.
- The secondary thread handles the encoding of frames received for preview.

### CODE REVIEW

File: src-tauri/src/capturer/mod.rs [hunk 1]
```diff
@@ -1,76 +1,61 @@
...
+
+mod encoder;
 
 fn get_random_id() -> String {
     let random_number: u64 = rand::thread_rng().gen();
     let id = format!("{:x}", random_number);
     id.chars().take(13).collect()
 }
 
+fn process_data(rx: mpsc::Receiver<Frame>, width: usize, height: usize, file_path: FileOutput) {
```
Comment (clearlysid) [inline src-tauri/src/capturer/mod.rs:26]: let's think of a better, clearer name for this function

File: src-tauri/src/capturer/mod.rs [hunk 2]
```diff
@@ -80,98 +65,108 @@ pub async fn start_capture(app_handle: AppHandle) {
     // record multiple gifs without restarting the app
     *frames = Vec::new();
```
Comment (clearlysid) [inline src-tauri/src/capturer/mod.rs:66]: I'm a little underconfident about this part. We call `start_recorder` first and then create a `Vec::new()` in `frames`. Are we sure this is the right order and works correctly? What happens if `start_recorder` has already captured a few frames by the time `Vec::new()` is called?

File: src-tauri/src/capturer/mod.rs [hunk 3]
```diff
@@ -80,98 +65,108 @@ pub async fn start_capture(app_handle: AppHandle) {
...
+    process_thread.join().expect("Processing thread panicked.");
 
+    println!("Creating Editor Window");
     println!("Preview path: {:?}", preview_path);
+    println!("Preview dimensions: {}x{}", output_width, output_height);
 
-    // Hide cropper, create editor
-    crate::cropper::toggle_cropper(&app_handle);
-    crate::toolbar::toggle_toolbar(&app_handle);
     crate::editor::init_editor(
```
Comment (clearlysid) [inline src-tauri/src/capturer/mod.rs:132]: At this moment in time when `init_editor()` is called, is the encoded video guaranteed to be ready? The code structure seems to suggest so.

Previously this was not true. Therefore, I fired a `preview-ready` event later (after creating editor window) when the encoding finished. Then the frontend knows that preview video is ready to load.

If the preview is guaranteed to be ready at the time of editor creation now, we can remove this event from both, the backend and frontend.

File: src-tauri/src/capturer/utils.rs
Comment (clearlysid) [inline src-tauri/src/capturer/utils.rs:1]: not happy with the title of this file tbh, we can maybe just call it `recorder` to be more explicit.

File: src-tauri/src/capturer/mod.rs (hunk 2 above)
Comment (anubhavitis) [reply src-tauri/src/capturer/mod.rs:66]: afaik, recording will be taken care by recorder.get_next_frame() func.

File: src-tauri/src/capturer/mod.rs (hunk 3 above)
Comment (anubhavitis) [reply src-tauri/src/capturer/mod.rs:132]: I have added the wait to make sure encoding is done.
```   
 process_thread.join().expect("Processing thread panicked.");
```

File: src-tauri/src/capturer/utils.rs
Comment (anubhavitis) [reply src-tauri/src/capturer/utils.rs:1]: utils.rs makes more sense since there are lot of util func implemented there.

**Comment by clearlysid** [comment]
closing in favour of #37

**Pull request closed without merging**